pkg/datapath/neighbor/desired_neighbor_calculator.go

```diff
@@ -0,0 +1,305 @@
+
+	// We always start out needing a full sync
+	needFullSyn := true
+
+	// Resync every 5 minutes, this will account of any deletions of forwardable IPs as well
+	// as catch any errors as result of an error in the partial sync.
+	const resyncInterval = 5 * time.Minute
+	resyncTimer := time.NewTimer(resyncInterval)
+
+	for {
+		rx := c.DB.ReadTxn()
+
+		// Get inserted or updated forwardable IPs since the last revision processed
+		var (
+			fipSeq   iter.Seq2[*ForwardableIP, statedb.Revision]
+			fipWatch <-chan struct{}
+		)
+
+		// If we need to do a full sync, we get all forwardable IPs
+		// Otherwise, we get only the ones that have been updated since the last revision
+		if needFullSyn {
+			fipSeq, fipWatch = c.ForwardableIPTable.AllWatch(rx)
+		} else {
+			fipSeq, fipWatch = c.ForwardableIPTable.LowerBoundWatch(rx, statedb.ByRevision[*ForwardableIP](fipRev))
+		}
+
+		// Get all L2 devices
+		l2Devs, l2DevWatch := l2Devices(c.DeviceTable, rx)
+
+		// Get all permutations of forwardable IPs and L2 devices
+		type fipWithDev struct {
+			fip *ForwardableIP
+			dev *tables.Device
+		}
+		permutations := iter.Seq[fipWithDev](func(yield func(fipWithDev) bool) {
+			for fip, rev := range fipSeq {
+				for _, dev := range l2Devs {
+					if !yield(fipWithDev{fip: fip, dev: dev}) {
+						// Store the last revision we processed
+						fipRev = rev
+						return
+					}
+				}
+			}
+		})
+
+		// Add a watch for the routes table as it was at the start of the
+		// transaction. We will be querying the FIB for next hops via netlink.
+		// We do this since netlink applies the exact same logic as the datapath does when
+		// determining the next hop for a given IP, which prevents subtle bugs.
+		// However, the answer we get will become invalid if the routes change.
+		// So we use the route table watch as signal for when to re-run this.
+		_, routeWatch := c.RouteTable.AllWatch(rx)
+
+		uniqueDesiredNeighbors := make(map[DesiredNeighborKey]struct{})
+
+		var errs error
+		// Find next hop routes for each forwardable IP
+		for fipDev := range permutations {
+			nextHop, err := c.getNextHopIP(fipDev.fip.IP, fipDev.dev.Index)
+			if err != nil {
+				if errors.Is(err, errNodeIPNotRoutable) {
+					// If the node IP is not routable, we don't need to do anything
+					continue
+				}
+
+				errs = errors.Join(errs, fmt.Errorf("failed to get next hop IP: %w", err))
+				continue
+			}
+
+			uniqueDesiredNeighbors[DesiredNeighborKey{
+				IP:      nextHop,
+				IfIndex: fipDev.dev.Index,
+			}] = struct{}{}
+		}
+
+		tx := c.DB.WriteTxn(c.DesiredNeighborTable)
+
+		// If we are doing a full sync, first delete all entries that should not be
+		// present anymore.
+		if needFullSyn {
+			for dn := range c.DesiredNeighborTable.All(rx) {
+				// If the desired neighbor is not in the set of unique desired neighbors,
+				// delete it.
+				if _, ok := uniqueDesiredNeighbors[dn.DesiredNeighborKey]; !ok {
+					_, _, err := c.DesiredNeighborTable.Delete(tx, dn)
```

Comment:
quick q: when a node in the cluster gets added and we add it to desired neighbor as well as a service backend, then that desired neighbor has two 'owners', what happens if one is going away but not the other? Do we have some implicit refcounting or detect this otherwise? (I presume agent restarts are survived as we persist the state table?)

Reply:
> I presume agent restarts are survived as we persist the state table

We do not persist the state table, we re-construct it after restart. The initializers ensure that we only start removing neighbors when we have fully rebuilt (all nodes and services processed)